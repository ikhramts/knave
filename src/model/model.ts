import { areArraysEqual } from '../utils';

export class KnaveModel {
    tables: Table[] = [];

    constructor(init?: Partial<KnaveModel>) {
        if (init) Object.assign(this, init);
    }

    table(name: string) : Table | null {
        let found = this.tables.filter(t => t.name == name);
        if (found.length > 0)
            return found[0];

        return null;
    }
}

export class Table {
    name: string;
    columns: Column[] = [];
    primaryKey: Index;

    constructor(init?: Partial<Table>) {
        if (init) Object.assign(this, init);
    }

    column(name: string) : Column{
        let found = this.columns.filter(c => c.name == name);
        if (found.length > 0)
            return found[0];

        return null;
    }
}

export class Column {
    name: string
    columnType: ColumnType;
    isNotNull: boolean;

    constructor(init?: Partial<Column>) {
        if (init) Object.assign(this, init);
    }
}

export class ColumnType {
    name: string;
    args: number[] = []

    constructor(init?: Partial<ColumnType>) {
        if (init) Object.assign(this, init);

        if (init.args) this.args = init.args.slice();
    }

    isEqualTo(other: ColumnType) : boolean {
        if (!other) return false;
        if (this.name != other.name) return false;
        if (!areArraysEqual(this.args, other.args)) return false;

        return true;
    }
}

export class Index {
    name: string;
    columns: Column[] = [];

    constructor(init?: Partial<Index>) {
        if (init) Object.assign(this, init);
    }
}