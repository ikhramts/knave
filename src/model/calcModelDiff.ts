import {
    KnaveModel, Table, Column, ColumnType, Index,
} from './model';
import {
    ModelDiff, AlteredTable, AlteredColumn, AlteredPrimaryKey
} from './modelDiff';

export function calcModelDiff(newModel: KnaveModel, oldModel: KnaveModel) : ModelDiff {
    // Do the diff on tables.
    let addedTables: Table[] = [];
    let removedTables: string[] = [];
    let alteredTables: AlteredTable[] = [];

    let newTablesByName = toMap(newModel.tables, table => table.name);
    let oldTablesByName = toMap(oldModel.tables, table => table.name);

    // Find added and altered tables.
    for (let newTable of newModel.tables) {
        let oldTable = oldTablesByName[newTable.name];

        if (oldTable) {
            // The table might have been altered.
            let alteredTable = calcTableDiff(newTable, oldTable);

            if (alteredTable) {
                alteredTables.push(alteredTable);
            } // Else: table was not altered.
        
        } else  {
            addedTables.push(newTable);
        }
    }

    // Find removed tables.
    for (let oldTable of oldModel.tables) {
        if (!newTablesByName[oldTable.name]) {
            removedTables.push(oldTable.name);
        }
    }

    let modelDiff = new ModelDiff();
    modelDiff.addedTables = addedTables;
    modelDiff.removedTables = removedTables;
    modelDiff.alteredTables = alteredTables;

    return modelDiff;
}

export function calcTableDiff(newTable: Table, oldTable: Table) : AlteredTable | null {
    if (newTable.name != oldTable.name) {
        throw "Column names are unexpectedly different. New column: " + newTable.name +
            "; old column: " + oldTable.name;
    }

    // Do diff on columns.
    let addedColumns: Column[] = [];
    let removedColumns: string[] = [];
    let alteredColumns: AlteredColumn[] = [];

    let newColumnsByName = toMap(newTable.columns, col => col.name);
    let oldColumnsByName = toMap(oldTable.columns, col => col.name);
    
    // Find added and altered columns.
    for (let newColumn of newTable.columns) {
        let colName = newColumn.name;
        let oldColumn = oldColumnsByName[colName];
        
        if (oldColumn) {
            // Might be altered
            let alteredColumn = calcColumnDiff(newColumn, oldColumn);

            if (alteredColumn) {
                alteredColumns.push(alteredColumn);
            } // Else: column was not altered.
        
        } else {
            addedColumns.push(newColumn)
        }
    }

    // Find removed columns.
    for (let oldColumn of oldTable.columns) {
        if (!newColumnsByName[oldColumn.name]) {
            removedColumns.push(oldColumn.name);
        }
    }

    // Find other differences.
    let alteredPrimaryKey = calcPrimaryKeyDiff(newTable.primaryKey, oldTable.primaryKey);
    
    // Compose output
    if (addedColumns.length == 0
        && removedColumns.length == 0
        && alteredColumns.length == 0
        && (!alteredPrimaryKey) ) 
    {
        return null;
    }

    let alteredTable = new AlteredTable();
    alteredTable.addedColumns = addedColumns;
    alteredTable.removedColumns = removedColumns;
    alteredTable.alteredColumns = alteredColumns;
    alteredTable.alteredPrimaryKey = alteredPrimaryKey;
    alteredTable.name = newTable.name;

    return alteredTable;
}

export function calcColumnDiff(newColumn: Column, oldColumn: Column) : AlteredColumn | null {
    if (newColumn.name != oldColumn.name) {
        throw "Column names are unexpectedly different. New column: " + newColumn.name +
            "; old column: " + oldColumn.name;
    }

    let newType = newColumn.columnType;
    let oldType = oldColumn.columnType;

    let columnsAreDifferent = 
        (newType.name != oldType.name) ||
        (!areArraysEqual(newType.args, oldType.args)) ||
        ((!newColumn.isNotNull) != (!oldColumn.isNotNull));

    if (columnsAreDifferent) {
        let alteredColumn = new AlteredColumn(newColumn);
        alteredColumn.originalColumn = oldColumn;
        return alteredColumn;
    }
    
    return null;
}

export function calcPrimaryKeyDiff(newIndex: Index, oldIndex: Index) : AlteredPrimaryKey | null {
    if ((!newIndex) && (!oldIndex)) {
        return null;
    }

    let alteredKey = new AlteredPrimaryKey();
    alteredKey.newIndex = newIndex;
    alteredKey.oldIndex = oldIndex;

    if ((!newIndex) || (!oldIndex)) return alteredKey;

    if (newIndex.name && newIndex.name != oldIndex.name) 
        return alteredKey;

    if (!containsSame(newIndex.columns, oldIndex.columns, index => index.name)) {
        return alteredKey;
    }
    
    return null;
}


function containsSame<T>(first: T[], second: T[], compareOn: (it:T) => string) : boolean {
    if ((!first) && (!second)) return true;
    if ((!first) || (!second)) return false;
    if (first.length != second.length) return false;
    
    let firstComparator = first.map(compareOn).sort();
    let secondComparator = second.map(compareOn).sort();
    let length = first.length;
    
    for (var i = 0; i < length; i++) {
        if (firstComparator[i] != secondComparator[i])
            return false;
    }

    return true;
}

function areArraysEqual<T>(first: T[], second: T[]) {
    if ((!first) && (!second)) return true;
    if ((!first) || (!second)) return false;
    if (first.length != second.length) return false;

    let length = first.length;

    for (let i = 0; i++; i < length) {
        if (first[i] != second[i])
            return false;
    }

    return true;
}

function toMap<T>(list: T[], getKey : (obj: T) => string) : {[id: string] : T} {
    if (!list) return null;
    if (list.length == 0) return {}

    let map : {[id: string] : T} = {};

    for (let item of list) {
        map[getKey(item)] = item;
    }

    return map;
}