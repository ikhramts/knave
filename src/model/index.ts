export class KnaveModel {
    tables: Table[] = [];

    constructor(init?: Partial<KnaveModel>) {
        if (init) (<any>Object).assign(this, init);
    }
}

export class Table {
    name: string;
    columns: Column[] = [];
    primaryKey: Index;

    constructor(init?: Partial<Table>) {
        if (init) (<any>Object).assign(this, init);
    }
}

export class Column {
    name: string
    columnType: ColumnType;
    isNotNull?: boolean;

    constructor(init?: Partial<Column>) {
        if (init) (<any>Object).assign(this, init);
    }
}

export class ColumnType {
    name: string;

    constructor(init?: Partial<ColumnType>) {
        if (init) (<any>Object).assign(this, init);
    }
}

export class Index {
    name: string;
    columns: Column[] = [];

    constructor(init?: Partial<Index>) {
        if (init) (<any>Object).assign(this, init);
    }
}