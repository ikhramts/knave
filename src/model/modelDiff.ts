import {
    Table,
    Column,
    ColumnType,
    Index
} from './model';

export class ModelDiff {
    addedTables: Table[];
    removedTables: string[];
    alteredTables: AlteredTable[];

    constructor(init?: Partial<ModelDiff>) {
        if (init) Object.assign(this, init);
    }
}

export class AlteredTable {
    originalTable: Table;

    addedColumns: Column[];
    removedColumns: string[];
    alteredColumns: AlteredColumn[];

    alteredPrimaryKey: Index | null;

    constructor(init?: Partial<AlteredTable>) {
        if (init) Object.assign(this, init);
    }
}

export class AlteredColumn extends Column {
    originalColumn: Column;

    constructor(init?: Partial<AlteredColumn>) {
        super({});
        if (init) Object.assign(this, init);
    }
}