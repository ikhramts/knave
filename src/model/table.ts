export class Table {
    name: string;

    constructor(init: Partial<Table>) {
        (Object as any).assign(this, init);
    }
}