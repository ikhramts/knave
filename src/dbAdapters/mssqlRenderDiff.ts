import * as model from '../model';

export function mssqlRenderDiff(modelDiff: model.ModelDiff) : string {
    let sql = "";

    for (let createTable of modelDiff.addedTables) {
        sql += renderCreateTable(createTable) + "\n";
    }

    for (let dropTable of modelDiff.removedTables) {
        sql += renderDropTable(dropTable);
    }

    for (let alterTable of modelDiff.alteredTables) {
        sql += renderAlterTable(alterTable);
    }

    return sql;
}

function renderDropTable(tableName: string): string {
    return 'DROP TABLE "' + tableName + '";\n';
}

function renderCreateTable(table: model.Table): string {
    let sql = 'CREATE TABLE "' + table.name + '" (\n    ';
    
    let columnsSql = table.columns.map(renderColumn);
    sql += columnsSql.join(',\n    ');

    let primaryKeySql = renderPrimaryKey(table.primaryKey, table.name);
    if (table.primaryKey) {
        sql += ',\n    ' + primaryKeySql;
    }

    sql += '\n);\n'

    return sql;
}

function renderColumn(column: model.Column) : string {
    let colType = column.columnType;
    let sql = '"' + column.name + '" ' + renderType(column.columnType);

    sql += ' ' + renderNullability(column.isNotNull);

    return sql;
}

function renderPrimaryKey(primaryKey: model.Index, tableName: string) : string {
    if (!primaryKey || !primaryKey.columns || primaryKey.columns.length == 0)
        return null;

    let pkeyName = getPrimaryKeyName(primaryKey.name, tableName);
    let sql = 'CONSTRAINT "' + pkeyName + '" PRIMARY KEY CLUSTERED (\n        ';

    let pkColumns = primaryKey.columns.map(c => '"' + c.name + '" ASC').join(',\n        ');
    sql += pkColumns + '\n    )';
    return sql;
}

function renderAlterTable(table: model.AlteredTable): string {
    let tableName = table.name;
    let sql = '--------------- ALTER TABLE "' + tableName + '" ---------------\n';

    let alterTable = () => 'ALTER TABLE "' + tableName + '" ';

    for (let addColumn of table.addedColumns) {
        sql += alterTable() + 'ADD ' + renderColumn(addColumn) + ';\n';
    }

    for (let dropColumn of table.removedColumns) {
        sql += alterTable() + 'DROP "' + dropColumn + '";\n';
    }

    for (let alterColumn of table.alteredColumns) {
        sql += alterTable() + renderAlterColumn(alterColumn);
    }

    if (table.alteredPrimaryKey) {
        let oldKey = table.alteredPrimaryKey.oldIndex;
        let newKey = table.alteredPrimaryKey.newIndex;

        if (oldKey) {
            let oldKeyName = getPrimaryKeyName(table.alteredPrimaryKey.oldIndex.name, table.name);
            sql += alterTable() + 'DROP CONSTRAINT "' + oldKeyName + '";\n';
        }

        if (newKey) {
            sql += alterTable() + 'ADD ' + renderPrimaryKey(newKey, table.name) + ';\n';
        }
    }

    sql += '\n';
    
    return sql;
}

function renderAlterColumn(alterColumn: model.AlteredColumn) : string {
    let sql = 'ALTER COLUMN "' + alterColumn.name + '"';

    if (!alterColumn.columnType.isEqualTo(alterColumn.originalColumn.columnType)) {
        sql += ' ' + renderType(alterColumn.columnType);
    }

    if((!alterColumn.isNotNull) != (!alterColumn.originalColumn.isNotNull)) {
        sql += ' ' + renderNullability(alterColumn.isNotNull);
    };

    return sql;
}

function renderType(colType: model.ColumnType) : string {
    let sql = colType.name;

    if (colType.args && colType.args.length > 0) {
        let argsSql = '(' + colType.args.map(arg => "" + arg).join(', ') + ')';
        sql += argsSql;
    }

    return sql;
} 

function renderNullability(isNotNull: boolean) : string {
    if (isNotNull) return 'NOT NULL';
    return 'NULL';
}

function getPrimaryKeyName(keyName: string, tableName: string) {
    if (keyName) return keyName;
    return "PK_" + tableName;
}

