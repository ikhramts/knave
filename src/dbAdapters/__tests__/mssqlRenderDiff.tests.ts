import * as model from '../../model';
import { mssqlRenderDiff } from '../mssqlRenderDiff';

describe("mssqlRenderDiff()", () => {
    it("Create table without primary key", () => {
        let modelDiff = new model.ModelDiff();
        modelDiff.addedTables.push(getTestTable());

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Create table with two columns", () => {
        let table = getTestTable();
        table.columns.push(getTestColumn('col2', 'type2'));
        let modelDiff = new model.ModelDiff();
        modelDiff.addedTables.push(table);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Create table with a primary key", () => {
        let table = getTestTable();
        let primaryKey = new model.Index();
        primaryKey.columns.push(table.columns[0]);
        table.primaryKey = primaryKey;

        let modelDiff = new model.ModelDiff();
        modelDiff.addedTables.push(table);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Create table with a primary key that references two columns", () => {
        let table = getTestTable();
        table.columns.push(getTestColumn('col2', 'type2'));

        let primaryKey = new model.Index();
        primaryKey.columns.push(table.columns[0]);
        primaryKey.columns.push(table.columns[1]);
        table.primaryKey = primaryKey;

        let modelDiff = new model.ModelDiff();
        modelDiff.addedTables.push(table);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Create table with primary key that has a name", () => {
        let table = getTestTable();
        let primaryKey = new model.Index({name: 'PK_someName'});
        primaryKey.columns.push(table.columns[0]);
        table.primaryKey = primaryKey;

        let modelDiff = new model.ModelDiff();
        modelDiff.addedTables.push(table);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Create table with null and not null columns", () => {
        let table = getTestTable();
        table.columns.push(getTestColumn('col2', 'type2'));
        table.columns[0].isNotNull = false;
        table.columns[1].isNotNull = true;

        let modelDiff = new model.ModelDiff();
        modelDiff.addedTables.push(table);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Create table with column that has type arguments", () => {
        let table = getTestTable();
        table.columns[0].columnType.args = [2, 3];
        let modelDiff = new model.ModelDiff();
        modelDiff.addedTables.push(table);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Create two tables", () => {
        let modelDiff = new model.ModelDiff();
        modelDiff.addedTables.push(getTestTable("table1"));
        modelDiff.addedTables.push(getTestTable("table2"));
        
        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Drop table", () => {
        let modelDiff = new model.ModelDiff();
        modelDiff.removedTables.push("table1");
        
        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Drop two tables", () => {
        let modelDiff = new model.ModelDiff();
        modelDiff.removedTables.push("table1");
        modelDiff.removedTables.push("table2");
        
        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Drop table and create table", () => {
        let modelDiff = new model.ModelDiff();
        modelDiff.addedTables.push(getTestTable("table1"));
        modelDiff.removedTables.push("table2");
        
        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Alter table - add column", () => {
        let alterTable = getTestAlterTable();
        alterTable.addedColumns.push(getTestColumn());

        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Alter table - add two columns", () => {
        let alterTable = getTestAlterTable();
        alterTable.addedColumns.push(getTestColumn("col1"));
        alterTable.addedColumns.push(getTestColumn("col2"));
        
        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Alter table - drop column", () => {
        let alterTable = getTestAlterTable();
        alterTable.removedColumns.push("col1");
        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Alter table - drop two columns", () => {
        let alterTable = getTestAlterTable();
        alterTable.removedColumns.push("col1");
        alterTable.removedColumns.push("col2");
        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Alter table - alter column - change type", () => {
        let alterColumn = getTestAlterColumn();
        alterColumn.columnType.name = 'type2';
        alterColumn.columnType.args = [3, 4];

        let alterTable = getTestAlterTable();
        alterTable.alteredColumns.push(alterColumn);
        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Alter table - alter column - change nullability", () => {
        let alterColumn = getTestAlterColumn();
        alterColumn.isNotNull = !alterColumn.originalColumn.isNotNull;
        
        let alterTable = getTestAlterTable();
        alterTable.alteredColumns.push(alterColumn);
        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Alter table - alter column - change type and nullability", () => {
        let alterColumn = getTestAlterColumn();
        alterColumn.isNotNull = !alterColumn.originalColumn.isNotNull;
        alterColumn.columnType.name = 'type2';
        
        let alterTable = getTestAlterTable();
        alterTable.alteredColumns.push(alterColumn);
        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Alter table - change primary key", () => {
        let alterKey = new model.AlteredPrimaryKey({ 
            oldIndex: getTestPrimaryKey(),
            newIndex: getTestPrimaryKey()
        });
        alterKey.newIndex.columns[0].name = 'col2';

        let alterTable = new model.AlteredTable({
            name: "table1", 
            alteredPrimaryKey: alterKey
        });
        
        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Alter table - remove primary key", () => {
        let alterKey = new model.AlteredPrimaryKey({ 
            oldIndex: getTestPrimaryKey(),
        });

        let alterTable = new model.AlteredTable({
            name: "table1", 
            alteredPrimaryKey: alterKey
        });
        
        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });

    it("Alter table - add primary key", () => {
        let alterKey = new model.AlteredPrimaryKey({ 
            newIndex: getTestPrimaryKey(),
        });

        let alterTable = new model.AlteredTable({
            name: "table1", 
            alteredPrimaryKey: alterKey
        });
        
        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();

    });

    it("Alter table - add primary key with custom name", () => {
        let alterKey = new model.AlteredPrimaryKey({ 
            newIndex: getTestPrimaryKey()
        });
        alterKey.newIndex.name = "CustomKeyName";

        let alterTable = new model.AlteredTable({
            name: "table1", 
            alteredPrimaryKey: alterKey
        });
        
        let modelDiff = new model.ModelDiff();
        modelDiff.alteredTables.push(alterTable);

        let sql = mssqlRenderDiff(modelDiff);
        expect(sql).toMatchSnapshot();
    });
});

function getTestTable(name?: string) : model.Table {
    if (!name) name = "table1";

    let column = getTestColumn();
    let table = new model.Table({
        name: name,
        columns: [column]
    });

    return table;
}

function getTestColumn(name?: string, typeName?: string) : model.Column {
    if (!name) name = "col1";
    if (!typeName) typeName = "type1";
    
    let colType = new model.ColumnType({name: typeName});
    let column = new model.Column({
        name: name,
        columnType: colType
    });
    
    return column;
}

function getTestAlterTable(name?: string) : model.AlteredTable {
    if (!name) name = "table1";
    
    let alterTable = new model.AlteredTable({name: name});
    return alterTable;
}

function getTestAlterColumn(name?: string, typeName?: string) : model.AlteredColumn {
    if (!name) name = "col1";
    if (!typeName) typeName = "type1";

    let originalColumn = getTestColumn(name, typeName);
    let alterColumnType = new model.ColumnType(originalColumn.columnType);
    let alterColumn = new model.AlteredColumn({
        originalColumn: originalColumn,
        name: name,
        columnType: alterColumnType
    });

    return alterColumn;
}

function getTestPrimaryKey(col1Name?: string) {
    if (!col1Name) col1Name = 'col1';

    let column = getTestColumn(col1Name);
    let primaryKey = new model.Index({columns: [column]});
    return primaryKey;
}

