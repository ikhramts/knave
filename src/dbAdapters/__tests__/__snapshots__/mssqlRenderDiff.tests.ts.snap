// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mssqlRenderDiff() Alter table - add column 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" ADD \\"col1\\" type1 NULL;

"
`;

exports[`mssqlRenderDiff() Alter table - add primary key 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" ADD CONSTRAINT \\"PK_table1\\" PRIMARY KEY CLUSTERED (
        \\"col1\\" ASC
    );

"
`;

exports[`mssqlRenderDiff() Alter table - add primary key with custom name 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" ADD CONSTRAINT \\"CustomKeyName\\" PRIMARY KEY CLUSTERED (
        \\"col1\\" ASC
    );

"
`;

exports[`mssqlRenderDiff() Alter table - add two columns 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" ADD \\"col1\\" type1 NULL;
ALTER TABLE \\"table1\\" ADD \\"col2\\" type1 NULL;

"
`;

exports[`mssqlRenderDiff() Alter table - alter column - change nullability 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" ALTER COLUMN \\"col1\\" NOT NULL
"
`;

exports[`mssqlRenderDiff() Alter table - alter column - change type 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" ALTER COLUMN \\"col1\\" type2(3, 4)
"
`;

exports[`mssqlRenderDiff() Alter table - alter column - change type and nullability 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" ALTER COLUMN \\"col1\\" type2 NOT NULL
"
`;

exports[`mssqlRenderDiff() Alter table - change primary key 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" DROP CONSTRAINT \\"PK_table1\\";
ALTER TABLE \\"table1\\" ADD CONSTRAINT \\"PK_table1\\" PRIMARY KEY CLUSTERED (
        \\"col2\\" ASC
    );

"
`;

exports[`mssqlRenderDiff() Alter table - drop column 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" DROP \\"col1\\";

"
`;

exports[`mssqlRenderDiff() Alter table - drop two columns 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" DROP \\"col1\\";
ALTER TABLE \\"table1\\" DROP \\"col2\\";

"
`;

exports[`mssqlRenderDiff() Alter table - remove primary key 1`] = `
"--------------- ALTER TABLE \\"table1\\" ---------------
ALTER TABLE \\"table1\\" DROP CONSTRAINT \\"PK_table1\\";

"
`;

exports[`mssqlRenderDiff() Create table with a primary key 1`] = `
"CREATE TABLE \\"table1\\" (
    \\"col1\\" type1 NULL,
    CONSTRAINT \\"PK_table1\\" PRIMARY KEY CLUSTERED (
        \\"col1\\" ASC
    )
);

"
`;

exports[`mssqlRenderDiff() Create table with a primary key that references two columns 1`] = `
"CREATE TABLE \\"table1\\" (
    \\"col1\\" type1 NULL,
    \\"col2\\" type2 NULL,
    CONSTRAINT \\"PK_table1\\" PRIMARY KEY CLUSTERED (
        \\"col1\\" ASC,
        \\"col2\\" ASC
    )
);

"
`;

exports[`mssqlRenderDiff() Create table with column that has type arguments 1`] = `
"CREATE TABLE \\"table1\\" (
    \\"col1\\" type1(2, 3) NULL
);

"
`;

exports[`mssqlRenderDiff() Create table with null and not null columns 1`] = `
"CREATE TABLE \\"table1\\" (
    \\"col1\\" type1 NULL,
    \\"col2\\" type2 NOT NULL
);

"
`;

exports[`mssqlRenderDiff() Create table with primary key that has a name 1`] = `
"CREATE TABLE \\"table1\\" (
    \\"col1\\" type1 NULL,
    CONSTRAINT \\"PK_someName\\" PRIMARY KEY CLUSTERED (
        \\"col1\\" ASC
    )
);

"
`;

exports[`mssqlRenderDiff() Create table with two columns 1`] = `
"CREATE TABLE \\"table1\\" (
    \\"col1\\" type1 NULL,
    \\"col2\\" type2 NULL
);

"
`;

exports[`mssqlRenderDiff() Create table without primary key 1`] = `
"CREATE TABLE \\"table1\\" (
    \\"col1\\" type1 NULL
);

"
`;

exports[`mssqlRenderDiff() Create two tables 1`] = `
"CREATE TABLE \\"table1\\" (
    \\"col1\\" type1 NULL
);

CREATE TABLE \\"table2\\" (
    \\"col1\\" type1 NULL
);

"
`;

exports[`mssqlRenderDiff() Drop table 1`] = `
"DROP TABLE \\"table1\\";
"
`;

exports[`mssqlRenderDiff() Drop table and create table 1`] = `
"CREATE TABLE \\"table1\\" (
    \\"col1\\" type1 NULL
);

DROP TABLE \\"table2\\";
"
`;

exports[`mssqlRenderDiff() Drop two tables 1`] = `
"DROP TABLE \\"table1\\";
DROP TABLE \\"table2\\";
"
`;
